cleverhans/future/torch/attacks/spsa.py

```diff
@@ -0,0 +1,187 @@
+            https://arxiv.org/abs/1611.01236). Default is None.
+  :param targeted: (optional) bool. Is the attack targeted or untargeted? Untargeted, the
+            default, will try to make the label incorrect. Targeted will instead try to
+            move in the direction of being more like y.
+  :param early_stop_loss_threshold: A float or None. If specified, the attack will end as
+            soon as the loss is below `early_stop_loss_threshold`.
+  :param learning_rate: Learning rate of ADAM optimizer.
+  :param delta: Perturbation size used for SPSA approximation.
+  :param spsa_samples:  Number of inputs to evaluate at a single time. The true batch size
+            (the number of evaluated inputs for each update) is `spsa_samples *
+            spsa_iters`
+  :param spsa_iters:  Number of model evaluations before performing an update, where each
+            evaluation is on `spsa_samples` different inputs.
+  :param is_debug: If True, print the adversarial loss after each update.
+  :param sanity_checks: bool, if True, include asserts (Turn them off to use less runtime /
+            memory or for unit tests that intentionally pass strange input)
+  :return: a tensor for the adversarial example
+  """
+
+  if y is not None and len(x) != len(y):
+    raise ValueError('number of inputs {} is different from number of labels {}'
+                     .format(len(x), len(y)))
+  if y is None:
+    y = torch.argmax(model_fn(x), dim=1)
+
+  # The rest of the function doesn't support batches of size greater than 1,
+  # so if the batch is bigger we split it up.
+  if len(x) != 1:
+    adv_x = []
+    for x_single, y_single in zip(x, y):
+      adv_x_single = spsa(model_fn=model_fn, x=x_single.unsqueeze(0), eps=eps,
+                          nb_iter=nb_iter, clip_min=clip_min, clip_max=clip_max,
+                          y=y_single.unsqueeze(0), targeted=targeted,
+                          early_stop_loss_threshold=early_stop_loss_threshold,
+                          learning_rate=learning_rate, delta=delta,
+                          spsa_samples=spsa_samples, spsa_iters=spsa_iters,
+                          is_debug=is_debug, sanity_checks=sanity_checks)
+      adv_x.append(adv_x_single)
+    return torch.cat(adv_x)
+
+  if eps < 0:
+    raise ValueError(
+        "eps must be greater than or equal to 0, got {} instead".format(eps))
+  if eps == 0:
+    return x
+
+  if clip_min is not None and clip_max is not None:
+    if clip_min > clip_max:
+      raise ValueError(
+          "clip_min must be less than or equal to clip_max, got clip_min={} and clip_max={}"
+          .format(clip_min, clip_max))
+
+  asserts = []
+
+  # If a data range was specified, check that the input was in that range
+  asserts.append(torch.all(x >= clip_min))
+  asserts.append(torch.all(x <= clip_max))
+
+  if is_debug:
+    print("Starting SPSA attack with eps = {}".format(eps))
+
+  perturbation = (torch.rand_like(x) * 2 - 1) * eps
```

Comment:
Does this imply the attack assumes that the inputs are centered around 0?

Reply:
No, because perturbation is eventually added to x (see line 113)